Adding the trajectories endpoint to OPTIMADE (#377)
Materials-Consortia/OPTIMADE: optimade.rst

    For example, a :entry:`structures` entry is comprised by data that belong to a single structure.

**Entry type**

>     Entries are categorized into types, e.g., :entry:`structures`, :entry:`trajectories`, :entry:`calculations`, :entry:`references`.

Reviewer: This is supposed to be a small set of examples, I do not think it needs more than 3 items.

Author: Ok, I will remove ":entry:`trajectories`, "

    This is particularly relevant for the default JSON-based response format.
    In this case, **field** refers to the name part of the name-value pairs of JSON objects.

**Trajectory**

>     A Trajectory contains data belonging to a set of structures.

Reviewer: I would replace "set" with "sequence" to make the relationship between structures clearer.

Author: I have replaced set with sequence for now. 
It depends a bit on how broad you see the scope of the trajectory endpoint. If you want to share a number of structures that are unordered, set would be more appropriate. 
There is/was also a proposal for a collections endpoint. If we accept that proposal, there will be an alternative way to group structures. In that case it won't be necessary to do this via the trajectory endpoint and this definition can be a bit more restrictive by using sequence instead of set.

Reviewer: > It depends a bit on how broad you see the scope of the trajectory endpoint. If you want to share a number of structures that are unordered, set would be more appropriate.

Then I would argue that name "trajectory" is no longer accurate, or am I missing something?

Author: No you're not missing anything. Trajectory would indeed not be such a good name in that case. So far, there are no plans to use the trajectory endpoint in such a broad sense. But I can't exclude the possibility that some future group will want to use it in such a way.

Reviewer 2: Isn't the design of the trajectory endpoint fundamentally relying on the assumption that it describes an *ordered* collection of frames - since they are indexed by a frame number; i.e., it represents them as a sequence rather than a set?

(And, if one really wants to represent a set of structures, shouldn't one use a #386 collection of structures instead?)

Author: Let's say, I have a training set for generating machine learned potentials. It would feature a set of structures with the same atoms but in different positions. The generated potential would not depend on the order of the structures, so there is no reason to place them in a specific order. So in this case the order of the structures would not be fundamental. 
Of course the order MUST be the same every time a request is made, but it does not matter in which order the structures are initially stored in the trajectory.
It would be more efficient to use the trajectory endpoint for sharing this data, because I would not need to repeat many of the fields like species, lattice_vectors, and species at sites. While this is not the purpose of the trajectory endpoint, it is possible it will be used like this in the future.

    Usually this data will come from molecular dynamics simulations.

>     It can however also contain data from structures that are related in an different way.

Reviewer: "an different" -> "a different"

    For example the successive structures from a Monte Carlo simulation.

**Frame**

>     An individual structure or data belonging to an individual structure from a trajectory.

Author: Matthew made the same comment and I have removed the definitions from 
Definition of Terms section and I have tried to add the same information to the description of the Trajectory endpoint.

> 

Reviewer 3: Discussion from the web meeting (28/09/22) was to move these definitions into the trajectory endpoint introduction

Author: Ok, I have removed these definitions here and added the same information to the description field of the trajectory endpoint.

A database provider MAY publish a special Index Meta-Database base URL. The main purpose of this base URL is to allow for automatic discoverability of all databases of the provider. Thus, it acts as a meta-database for the database provider's implementation(s).

The index meta-database MUST only provide the :endpoint:`info` and :endpoint:`links` endpoints, see sections `Info Endpoints`_ and `Links Endpoint`_.

> It MUST NOT expose any entry listing endpoints (e.g., :endpoint:`structures` and :endpoint:`trajectories`).

Reviewer: Also an example here, no need to include `trajectories`.

Author: Ok, I will remove this.

  property, thus reducing the amount of data transferred and avoiding to perform separate HTTP requests for each
  individual property.

Other compact formats might be defined in future versions of this specification (e.g., constant values only in a
range of indices, or linearly varying values), and an appropriate way to communicate the format of the
compact representation will be defined.

The client MUST be able to handle responses with both compact and non-compact property formats.

Examples:

- A trajectory with a fixed unit cell for all frames could be represented as:

  .. code:: jsonc

    {
      "data":{
        "id": "traj00000001",
        "type": "trajectories",
        "attributes": {
          "nframes": 5,
          "last_modified":"2021-07-16T18:02:03Z",
          "elements": [["H","O"]],
          "nelements": [2],
          "lattice_vectors" : [
            [[4.0, 0.0, 0.0],[0.0, 4.0, 0.0],[0.0, 0.0, 4.0]]
          ],
          "_exmpl_timestep": [0.0, 1.0, 2.0, 3.0, 4.0],
          "cartesian_site_positions" : null,
          // ...
        },
      //...
    }

>   In this example, the properties :field:`elements`, :field:`nelements`, and :field:`lattice_vectors` are compacted along the first dimension (`dim_frames`).

Reviewer 4: Indeed but it's not part of the response, it's in the property definition.

  - A structure having implicit atoms and using assemblies: :val:`["assemblies", "implicit_atoms"]`

Trajectories Entries
--------------------

> - **Description**: The :entry:`trajectories` entry point contains data belonging to a set of structures. These structures are related because they were, for example, created by the same procedure, e.g. molecular dynamic trajectories, relaxations of a molecule or crystal structure, Monte Carlo simulations, etc. Some examples of the data that can be shared are the particle positions, the pressure and the energies. :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the following properties `reference\_structure`_, `reference\_frame`_, `nframes`_, `available_properties`_ and `next_part_trajectory`_. Next to this they can optionally have all the fields of the structures entries, including relationships, as well as database specific fields.

Reviewer 3: Sorry to be a pain, but could you please reformat this so that each sentence is on a new line (which will not incur a linebreak in rst)? This makes it easier for people to review/suggest changes on specific parts. I think we have tried to follow this for the rest of the spec. Other purely visual changes can wait until the content has been agreed upon.

Author: I have placed every sentence on a separate line.

  The `reference\_structure`_ is an example of the kind of structures that are in the trajectory, and it is used to query the trajectory entries in the same way as the structures entries. The data belonging to the frames of the trajectory is only returned when this is specifically requested in the :query-param:`response_fields` parameter. In this case each property has a dictionary as a value which contains the values of this property and information about which value belongs to which frame. It is possible to request only part of a trajectory and to request only 1 out of every n frames. Queries on individual frames are currently not supported.

reference\_structure

> ~~~~~~~~~~~~~~~~~~~~

Reviewer 3: I think my suggestion would be to include the `reference_structure` as a mandatory `relationship` for each trajectory, so that filtering can be enabled as a relationship filter (which will probably have to become mandatory too), e.g. `/trajectories?filter=structure.elements HAS "C"`. This feels like the most JSON:API-esque way of doing this.

Author: I think the JSON relationship field should only be used for references. 
So if Iunderstand it correctly you are suggesting to make the reference structure a structure like any other structure in the structures endpoint, and that the reference structure field, within the relationship section of a trajectory, would have a link to this structure.

If I remeber correctly, during the Optimade meeting the point was raised that a trajectory should have all the data required to plot a trajectory. And as much of the data in the reference structure is needed to plot the trajectory it seemed logical to place it in the trajectory. Later we made it so practically any field can be defined on a per frame basis so now it may make more sense to have a seperate structure as the reference structure. 

One other point that we have to think of, if the reference structures are turned into normal structure entry point structures, is that the structures that are currently in the databases are all optimized structures. This may not be the case for a reference structure. We should somehow make sure that for someone who searches through the structures it is clear wether or not the structure is optimized.

Author: I recently spoke with Josh Vita from Open KIM. They use machine learning to try to generate interatomic potentials based on quantum mechanical simulations. They have training sets with many structures. We discussed whether they could use the trajectory end points for sharing these training sets. The individual structures in these training sets can, however, have different elements. So there is not a clear way for them to define a consistent reference structure. For example, should the elements field hold all the elements in the trajectory or just the elements for which the structure is given?

One option could be to have a set of reference properties that apply to all the frames and a separate optional reference structure. I am curious to hear what your opinions on this are.

Author: I am closing this discussion. If I remember correctly, the discussions did not lead to a decision to change this, and it has also been open for a very long time without replies. Feel free to reopen it though.

  - **Query**: MUST be a queryable property with support for all mandatory filter features.
  - The integer value MUST be equal to the length of the trajectory, that is, the number of frames.
  - The integer MUST be a positive non-zero value.

- **Querying**:

  - A filter that matches trajectories that have exactly 100 frames:
    - :filter:`nframes=100`
  - A filter that matches trajectories that have between 100 and 1000 frames:
    - :filter:`nframes>=100 AND nframes<=1000`

- **Examples**:

  -   :val:`42`

available_properties
~~~~~~~~~~~~~~~~~~~~

- **Description**: A list of the names of the properties for which data is available in the trajectory. It is up to the server to decide which properties to share and there are no mandatory fields. When sharing `cartesian\_site\_positions`_ the lattice_vectors, species, dimension_types and species_at_sites MUST be shared as well .
- **Type**: List of strings
- **Requirements/Conventions**:

  -   **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  -   **Query**: MUST be a queryable property with support for all mandatory filter features.

- **Examples**:

>   - :val:`["cartesian\_site\_positions","lattice_vectors","species",species_at_sites","_exmpl_Pressure"]`

Reviewer 3: All field names need to be lower case

```suggestion
  - :val:`["cartesian_site_positions","lattice_vectors","species",species_at_sites","_exmpl_pressure"]`
```

Author: Thanks for pointing this out, I will change it in the next commit.

next_part_trajectory
~~~~~~~~~~~~~~~~~~~~

> - **Description**: This is a link the client can use to download the remainder of the trajectory data when the server returns fewer frames than the client requested.

Reviewer 3: Could you explain how this is different to pagination in JSON:API? Could we not use the same mechanism, e.g. [a `links` field for the trajectory entry itself](https://jsonapi.org/format/#document-resource-object-links)?

Author: In principle, it should be possible to use the links field.
It would however make creating the next link a bit more complicated. (assuming the query would return multiple trajectories.)

You would have to specify what the original query was,
Next, you would need to specify which trajectory you are in the process of returning.
And finally, you would have to specify what is the first unreturned frame of that trajectory. 

As these extra fields should only be encountered by the server implementation that created the "next" link. We would not have to define these fields. So it would actually make writing the standard a bit easier. 

If you think this is a good idea, I can change the draft specification.

Reviewer 3: If the user wants to download an entire trajectory/more of a specific trajectory, would they not go to the single trajectory endpoint? In which case a self-link could be used (see e.g. optional fields in the section [Entry Listing JSON Response Schema](https://github.com/Materials-Consortia/OPTiMaDe/blob/develop/optimade.rst#532entry-listing-json-response-schema)).

It's not clear to me what the intention of `next_part_trajectory` is from what is written here, what will the user receive if they follow that link?

Author: Via the link in `next_part_trajectory` The user would receive the next part of the trajectory. Trajectories may be very large, so you would want to have the option for the server to return it one block at a time. 

In principle, the client can make a query that would return multiple complete trajectories by combining a filter with the appropriate response fields. In that case, I was originally thinking that the `next` field in the links section could point to the next trajectory, while the "next_part_trajectory" would point to the next block of the current trajectory.

Author: For now, I have removed the next_part_trajectory field.

- **Type**: Link
- **Requirements/Conventions**:

  - **Support**: MUST be present when the number of returned frames is smaller than the number of requested frames. MUST NOT be present otherwise.
- **Examples**:
    If, for example, the server has returned data on the first 30 frames it could provide this link to the client to download the next part of the trajectory.

  - :query-url:`http://example.com/optimade/v1/trajectories/traj00000001?response\_fields=cartesian_site_positions, lattice_vectors,dimension_types,species,species_at_sites&first_frame=30`

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The preceding properties 7.3.1-4 and the fields described under `Properties Used by Multiple Entry Types`_ MUST be returned when no response\_fields property (see the section `Entry Listing URL Query Parameters`_) is specified.

Author: If anyone knows how to reference a section by number such that the section number is automatically updated, I would gladly hear it. Writing out all the subsection names would be the alternative, but I think this would make this sentence very long.

- **frame_encoding**:

  - **Description**: To improve the compactness of the data there are several ways to show to which frame a value belongs. This is specified by the frame encoding parameter.
  - **Type**: string
  - **Requirements/Conventions**: This field MUST be present.
  - **Values**:

    - **constant**: The value of the property is constant and thus has the same value for each frame in the trajectory.
    - **explicit**: A value is given for each frame. The number of values MUST thus be equal to the number of frames and MUST be in the same order as the frames. If there is no value for a particular frame the value MUST be :val:`null`.
    - **linear**: The value is a linear function of the frame number. This function is defined by offset_linear and step_size_linear.
    - **explicit_regular_sparse**: The value is set every one per **step_size_sparse** frames, with **offset_sparse** as the first frame.

>     - **explicit_custom_sparse**: A separate list with frame numbers is defined in the field frame_number to indicate to which frame a value belongs.

Author: I have listed all the options that we discussed during the meeting. We also considered that perhaps  having all these 5 options would be a bit too much, but we did not decide which ones would not be worth implementing.

Reviewer 4: Let's discuss - it's probably OK to keep all, even if the clients must be a bit more complex - but maybe better to keep now rather than to add later, that will mean old clients will not work?

Author: Ok, I will keep all the options and mark this as resolved. Feel free to reopen it if you want to discuss it further.

  - **Type**: List of integers
  - **Requirements/Conventions**: The value MUST be present when frame_encoding is set to explicit_custom_sparse. Otherwise it MUST NOT be present. The frame numbers MUST be in the same order as the values.
  - **Examples**:

    - :val:`[0,20,78,345]`

- **values**:

  - **Description**: The values belonging to this property. The format of this field depends on the property and on  the frame_encoding parameter.
  - **Type**: Any
  - **Requirements/Conventions**: The value MUST be present. If a value has not been sampled for a particular frame the value should be set to :val:`null` at the highest possible nesting level. In case of `cartesian_site_positions`_, a site that has the value :val:`null` for the x,y and z coordinates means that the site is not in the simulation volume. This may be useful for grand canonical simulations where the number of particles in the simulation volume is not constant.

Example of returned trajectory
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is an example of a JSON object that could be returned after the following query:
:query-url:`http://example.com/optimade/v1/trajectories/traj00000001`

> .. code:: jsonc
> 
>     {
>       "reference_structure":{
>         "elements": ["H","O"],
>         "nelements": 2,
>         "elements_ratios": [0.666667,0.333333],
>         "chemical_formula_descriptive": "H2O",
>         "chemical_formula_reduced": "H2O",
>         "chemical_formula_anonymous": "A2B",
>         "dimension_types":[0,0,0],
>         "nperiodic_dimensions": 0,
>         "lattice_vectors" : [[4.0,0.0,0.0],[0.0,4.0,0.0],[0.0,0.0,4.0]],
>         "cartesian_site_positions" : [[2.0,2.0,2.0],[1.238,2.0,1.416],[2.762,2.0,1.416]],
>         "nsites":3,
>         "species_at_sites":["O1","H1","H2"],
>         "species":[
>           {
>             "name":"O1",
>             "chemical_symbols":["O"],
>             "concentration":[1.0],
>           },{
>             "name":"H1",
>             "chemical_symbols":["H"],
>             "concentration":[1.0],
>           },{
>             "name":"H2",
>             "chemical_symbols":["H"],
>             "concentration":[1.0],
>           }
>         ]
>       },
>       "id": "traj00000001",
>       "type": "trajectories",
>       "last_modified":"2021-07-16T18:02:03Z"
>       "reference_frame": 359,
>       "nframes": 360
>       "relationships": {
>         "references": {
>           "data": [
>             {
>               "type": "references",
>               "id": "dummy/2019"
>             }
>           ]
>         }
>       }
>       "available_properties":[
>         "cartesian_site_positions",
>         "species",
>         "dimension_types",
>         "lattice_vectors",
>         "species_at_sites",
>         "_exmpl_temperature_set",
>         "_exmpl_time",
>         "_exmpl_ekin"
>       ]
>     }

Reviewer 3: This is not a valid JSON:API response, this should be nested under a `"data"` field and all OPTIMADE defined fields must be nested under `"attributes"`. I've sketched this out below but it is difficult to format as a GitHub suggestion

```suggestion
.. code:: jsonc

    "data": [
      {          
         "id": "traj00000001",
         "type": "trajectories",
         "attributes": {
           "reference_structure":{
             "elements": ["H","O"],
             "nelements": 2,
             "elements_ratios": [0.666667,0.333333],
             "chemical_formula_descriptive": "H2O",
             "chemical_formula_reduced": "H2O",
             "chemical_formula_anonymous": "A2B",
             "dimension_types":[0,0,0],
             "nperiodic_dimensions": 0,
             "lattice_vectors" : [[4.0,0.0,0.0],[0.0,4.0,0.0],[0.0,0.0,4.0]],
             "cartesian_site_positions" : [[2.0,2.0,2.0],[1.238,2.0,1.416],[2.762,2.0,1.416]],
             "nsites":3,
             "species_at_sites":["O1","H1","H2"],
             "species":[
               {
                 "name":"O1",
                 "chemical_symbols":["O"],
                 "concentration":[1.0],
               },
               {
                 "name":"H1",
                 "chemical_symbols":["H"],
                 "concentration":[1.0],
               },
               {
                 "name":"H2",
                 "chemical_symbols":["H"],
                 "concentration":[1.0],
               }
             ]
          },
          "last_modified":"2021-07-16T18:02:03Z"
          "reference_frame": 359,
          "nframes": 360,
          "available_properties":[
            "cartesian_site_positions",
            "species",
            "dimension_types",
            "lattice_vectors",
            "species_at_sites",
            "_exmpl_temperature_set",
            "_exmpl_time",
            "_exmpl_ekin"
          ]
      },
      "relationships": {
        "references": {
          "data": [
            {
              "type": "references",
              "id": "dummy/2019"
            }
          ]
        }
      }
      
    },
    ...
    ]
```

Author: I have tried to correct this in the latest commit. I am however not entirely sure it is correct so please check it again,

After the previous querry is an example of a JSON object that could be returned after the following query:
:query-url:`http://example.com/optimade/v1/trajectories/traj00000001?response\_fields=cartesian_site_positions, lattice_vectors,dimension_types,_exmpl_time,_exmpl_ekin,species,species_at_sites,relationships&first_frame=0&frame_step=5`

> .. code:: jsonc

Reviewer 3: Same comment as above, this needs `"data"`, `"attributes"` etc to be valid JSON:API

  - A structure having implicit atoms and using assemblies: :val:`["assemblies", "implicit_atoms"]`

Trajectories Entries
--------------------

- **Description**: The :entry:`trajectories` entry point contains data belonging to a set of structures. These structures are related because they were, for example, created by the same procedure, e.g. molecular dynamic trajectories, relaxations of a molecule or crystal structure, Monte Carlo simulations, etc. Some examples of the data that can be shared are the particle positions, the pressure and the energies. :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the following properties `reference\_structure`_, `reference\_frame`_, `nframes`_, `available_properties`_ and `next_part_trajectory`_. Next to this they can optionally have all the fields of the structures entries, including relationships, as well as database specific fields.

>   The `reference\_structure`_ is an example of the kind of structures that are in the trajectory, and it is used to query the trajectory entries in the same way as the structures entries. The data belonging to the frames of the trajectory is only returned when this is specifically requested in the :query-param:`response_fields` parameter. In this case each property has a dictionary as a value which contains the values of this property and information about which value belongs to which frame. It is possible to request only part of a trajectory and to request only 1 out of every n frames. Queries on individual frames are currently not supported.

Author: As I stated at the end of this line, I have not defined anything about querying the individual frames. I did not include it because I think that, when such queries are done on many trajectories, it would take a long time for the server to process the query. Do you agree with this decision, or do you think we should define a way to query the whole trajectory?

  It can be used to select trajectories with queries and to give a quick visual impression of the structures in this trajectory.
- **Type**: dictionary
- **Requirements/Conventions**:

  - Each trajectory MUST have a reference_structure.
  - This reference_structure MAY be one of the frames from the trajectory, in that case the `reference\_frame`_ field MUST specify which frame has been used.
  - Queries on the trajectories MUST be done on the information supplied in the reference\_structure when the queried property is in the reference\_structure.
    The subfields of the reference\_structure MUST have the same queryability as in the `structures entries`_.

  - This reference frame has the same properties as the structure entries namely:

    - `elements`_
    - `nelements`_
    - `elements\_ratios`_
    - `chemical\_formula\_descriptive`_
    - `chemical\_formula\_reduced`_
    - `chemical\_formula\_hill`_
    - `chemical\_formula\_anonymous`_
    - `dimension\_types`_
    - `nperiodic\_dimensions`_
    - `lattice\_vectors`_
    - `cartesian\_site\_positions`_
    - `nsites`_
    - `species\_at\_sites`_
    - `species`_
    - `assemblies`_
    - `structure\_features`_

>   - It MAY also have a 'relationships'_ field for references that apply to the entire trajectory rather than to a single frame.

Author: During the discussions, it was mentioned that it would be nice if a certain frame could be linked to a structure entry with the same structure. My first idea was to treat the relationships field in the same way as the other properties that have a value for a frame.
I however also wanted to have the option to have references relating to the entire trajectory and not to a single frame, and therefore I placed an extra relationships section in the reference structure.

I am now however starting to doubt whether this is such a good idea, especially because the JSON API standard also defines a relationships field. And the way I am currently using the relationship fields is probably not what the makers of the JSON API standard intended. 

So my main question is: Do you think it is necessary to be able to make references from a frame?
If not, I can simply remove the relationships field from reference_structure and use one global  relationships field.

Reviewer 3: Just to clarify, the relationships field in the specification *is* the relationships field from JSON:API, it's just a general way of providing mappings from one entry to another, irrespective of their type.

You are right that JSON:API has no way of specifying relationships *within* an entry (e.g. individual relationships for frames within an entry types trajectory), and similarly listing them all in the top-level relationships is probably not a good idea.

Again, I think this depends on the intended use-case. Relationships don't have to be bi-directional, so you could have a one-to-one structures->trajectory relationship without the one-to-many trajectory->structure relationship. This would still allow you to grab all structures registered to a trajectory with something like `/structures?filter=trajectories.id=12345` but would not allow the reverse, e.g. `/trajectories?filter=structures.id=67890`.

Another problem with using relationships here is that JSON:API requires that you provide the linked documents in an `included` field if requested, which could become very large in both cases (would need very strict response field trimming, which will probably be the case for trajectories anyway).

Author: It is not mandatory to support the `included` field. If it is supported, the server would only have to return the mandatory fields of the trajectory. i.e. the reference frame and a few other fields. This would not be significantly more data than a typical structure entry.

- **Type**: integer
- **Requirements/Conventions**:

  - **Support**: MUST be supported if the reference\_structure is taken from the trajectory.
    If the reference\_structure is not in the trajectory, the value MUST NOT be present.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.

- **Examples**:

  - :val:`42`

nframes
~~~~~~~

- **Description**: The number of the frames in the trajectory.
- **Type**: integer
- **Requirements/Conventions**:

  - **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  - **Query**: MUST be a queryable property with support for all mandatory filter features.
  - The integer value MUST be equal to the length of the trajectory, that is, the number of frames.
  - The integer MUST be a positive non-zero value.

- **Querying**:

  - A filter that matches trajectories that have exactly 100 frames:
    - :filter:`nframes=100`
  - A filter that matches trajectories that have between 100 and 1000 frames:
    - :filter:`nframes>=100 AND nframes<=1000`

- **Examples**:

  -   :val:`42`

> available_properties

Author: At present, this field only stores data about which properties are available. Do we also want to include information about how many values there are for each property, or which frame encoding the property has, so it may be easier to preallocate memory?

Reviewer 4: This might be indeed useful (I think it should always be easy to compute on the server side?)

Author: I have added the `frame_serialization_format` and the number of values to the `available_properties` field.
This should help the client to estimate the amount of data that will be returned.

- **Type**: dictionary
- **Requirements/Conventions**:

  - Each trajectory MUST have a reference_structure.
  - This reference_structure MAY be one of the frames from the trajectory, in that case the `reference_frame`_ field MUST specify which frame has been used.
  - Queries on the trajectories MUST be done on the information supplied in the reference_structure when the queried property is in the reference_structure.
    The subfields of the reference_structure MUST have the same queryability as in the `structures entries`_.

  - This reference frame has the same properties as the structure entries namely:

    - `elements`_
    - `nelements`_
    - `elements_ratios`_
    - `chemical_formula_descriptive`_
    - `chemical_formula_reduced`_
    - `chemical_formula_hill`_
    - `chemical_formula_anonymous`_
    - `dimension_types`_
    - `nperiodic_dimensions`_
    - `lattice_vectors`_
    - `cartesian_site_positions`_
    - `nsites`_
    - `species_at_sites`_
    - `species`_
    - `assemblies`_
    - `structure_features`_

reference_frame
~~~~~~~~~~~~~~~~

- **Description**: The number of the frame at which the reference_structure was taken.
- **Type**: integer

> - **Requirements/Conventions**:

Reviewer 4: Specify also here that this is zero-based, and the value must be >=0 and < nframes

Author: I will expand the description for the reference frame.

  -   **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  -   **Query**: MUST be a queryable property with support for all mandatory filter features.

- **Examples**:

  - :val:`["cartesian_site_positions","lattice_vectors","species","dimension_types","species_at_sites","_exmpl_pressure"]`

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The preceding properties 7.3.1-4 and the fields described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.

The data from the trajectory frames SHOULD only be returned when the user specifically requests these properties in the response_fields.

Next to this the client MAY specify the following parameters to customize the return from the server.
While these URL query parameters are OPTIONAL for clients, API implementations MUST accept and handle them.
The numbering of the frames is zero based, so the first frame is frame number 0.

- **first_frame**:

  - **Description**: **first_frame** specifies the first frame that should be returned.
  - **Type**: integer

>   - **Requirements/Conventions**: The value MUST be larger or equal to 0 and MUST be less than or equal to nframes.(The total number of frames in the trajectory)

Reviewer 4: ```suggestion
  - **Requirements/Conventions**: The value MUST be larger or equal to 0 and MUST be less than nframes.(The total number of frames in the trajectory)
```

    If this is not the case :http-error:`400 Bad Request` MUST be returned with a message indicating that the value for this field is incorrect.
    The default value is 0.
  - **Examples**:

    - :query-url:`/trajectories/traj00000001?first_frame=1000`

- **last_frame**:

  - **Description**: **last_frame** specifies the last frame that should be returned.
  - **Type**: integer

>   - **Requirements/Conventions**: The value MUST be larger or equal to first_frame and MUST be less than or equal to nframes(the total number of frames in the trajectory).

Reviewer 4: ```suggestion
  - **Requirements/Conventions**: The value MUST be larger or equal to first_frame and MUST be less than nframes(the total number of frames in the trajectory).
```

    If this is not the case :http-error:`400 Bad Request` MUST be returned with a message indicating that the value for this field is incorrect.
    The default value is nframes.
  - **Examples**:

    - :query-url:`/trajectories/traj00000001?last_frame=2000`

- **frame_step**:

  - **Description**:  Specifies that data should only be returned for one out of every **frame_step** frames.
  - **Type**: integer

>   - **Requirements/Conventions**: The value MUST be larger or equal to 1 and MUST be less than or equal to the total number of frames.

Reviewer 4: ```suggestion
  - **Requirements/Conventions**: The value MUST be larger than or equal to 1.
```
(I suggest removing the limitation on the max value - if it's too large, you will just get one frame, but I think this is OK). We might actually think at whether to reduce the contraints on  first and last frame (keeping >=0, but removing the largest value: e.g. like in python, if `a=[1,2,3]` you can do `a[7:12]` and this just returns an empty list.)

Author: I think we can indeed leave this out. The server code would have to handle the case where the remainder is not dividable by the frame_step anyway.

    If this is not the case :http-error:`400 Bad Request` MUST be returned with a message indicating that the value for this field is incorrect.
    The default value is 1.
  - **Examples**:

    - :query-url:`/trajectories/traj00000001?frame_step=10`.

The server can decide how many frames are returned for each request.
If the number of frames returned is less than the total number of frames requested by a client the server has to supply a link in the :field:`next` field of the :field:`links` section of the JSON response, which the client can use to download the rest of the trajectory.

Return format for Trajectory Data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The returned data is first grouped per property and then by frame.
The property can be any of the fields described under `structures entries`_ or a database specific field.
Each property has a dictionary as the value, with the following fields:

> - **frame_encoding**:

Reviewer 4: I think the name 'encoding' is, in the end, correct in meaning - but I'm wondering if this might become confusing as people typically think to "string encoding" when "encoding" is mentioned? I don't have a better suggestion though (maybe "frame_serialization_format" but it's a bit long)

Author: I have changed it to frame_serialization_format for now. If you still come up with a better name i can easily change it with a search and replace.

  - **Values**:

    - **constant**: The value of the property is constant and thus has the same value for each frame in the trajectory.
    - **explicit**: A value is given for each frame.
      The number of values MUST thus be equal to the number of frames and MUST be in the same order as the frames.
      If there is no value for a particular frame the value MUST be :val:`null`.
    - **linear**: The value is a linear function of the frame number.
      This function is defined by offset_linear and step_size_linear.
    - **explicit_regular_sparse**: The value is set every one per **step_size_sparse** frames, with **offset_sparse** as the first frame.
    - **explicit_custom_sparse**: A separate list with frame numbers is defined in the field frame_number to indicate to which frame a value belongs.

- **offset_linear**:

  - **Description**: If **frame_encoding** is set to "linear" this property gives the value at frame 0.
  - **Type**: float
  - **Requirements/Conventions**: The value MAY be present when **frame_encoding** is set to "linear", otherwise the value MUST NOT be present.
    The default value is 0.
  - **Examples**:

    - :val:`1.5`

- **step_size_linear**:

>   - **Description**: If **frame_encoding** is set to "linear", this value gives the increase/decrease in the value of the property when going from one frame to the next.

Reviewer 4: Maybe clarify even more that this is the value change for a change by one unit of the frame number - even if some properties might be defined less often? (If you feel that's already clear enough, then no need to change)

Author: I will add a small example, so it will hopefully be a bit clearer to the reader.

  - **Type**: float
  - **Requirements/Conventions**: The value MUST be present when **frame_encoding** is set to "linear".
    Otherwise it MUST NOT be present.
  - **Examples**:

    - :val:`0.0005`

- **offset_sparse**:

  - **Description**: If **frame_encoding** is set to "explicit_regular_sparse" this property gives the frame number to which the first value belongs.
  - **Type**: integer
  - **Requirements/Conventions**: The value MAY be present when **frame_encoding** is set to "explicit_regular_sparse", otherwise the value MUST NOT be present.
    The default value is 0.
  - **Examples**:

    - :val:`100`

- **step_size_sparse**:

  - **Description**: If **frame_encoding** is set to "explicit_regular_sparse", this value indicates that every step_size_sparse frames a value is defined.
  - **Type**: integer
  - **Requirements/Conventions**: The value MUST be present when frame_encoding is set to "explicit_regular_sparse".
    Otherwise it MUST NOT be present.
  - **Examples**:

    - :val:`100`

> - **frame_number**:

Reviewer 4: Maybe rename to `sparse_frame_numbers` that is, I think, more clear than `frame_number` (that to see seems too general and potentially confusing to users who don't read the specs). Also, update accordingly the example below (where it says "frames" instead).

Also: do you want to mention also here that the values will be `0<=x<nframes` for all elements `x` of `sparse_frame_numbers`?

Author: I will change it to `sparse_frames` that is a bit shorter and should also avoid any ambiguity.

  - **Description**: If **frame_encoding** is set to "explicit_custom_sparse", this field holds the frames to which the values in the value field belong.
  - **Type**: List of integers
  - **Requirements/Conventions**: The value MUST be present when **frame_encoding** is set to "explicit_custom_sparse".
    Otherwise it MUST NOT be present.
    The frame numbers MUST be in the same order as the values.
  - **Examples**:

    - :val:`[0,20,78,345]`

> - **values**:

Reviewer 4: Here (and above) I just have two related general questions, of which I'm not sure.

- is it OK to have some values specified only depending on the value of the string "encoding"? Probably yes and this can be written as a validation rule, but just to double check
- is it OK that the value depends on the type (I'm thinking also from a validation point of view)? Probably yes as it's always a list of values, just the length (and the meaning of which frames it's associated to) changes, but I wanted to double check

Author: > is it OK to have some values specified only depending on the value of the string "encoding"? Probably yes and this can be written as a validation rule, but just to double check

I think this situation is quite similar to the structure_features field. 
If the string "assemblies" is present within `structure_features`  the field `assemblies` must be present as well.
If "assemblies" is not present the `assemblies` field should be `null` or not present.

> is it OK that the value depends on the type (I'm thinking also from a validation point of view)? Probably yes as it's always a list of values, just the length (and the meaning of which frames it's associated to) changes, but I wanted to double check

I think it is OK, At the info/structures endpoint the data type of all the properties in the database is listed. So we could check whether the data type of the items in the values field matches the type specified there. For the properties in the Optimade standard there is also the openapi_schema.json, where the data type of all OPTIMADE defined properties is listed.  Most of the validator was written by @ml-evs and @CasperWA though, so they can probably give a more definitive answer.

  - **Description**: The values belonging to this property.
    The format of this field depends on the property and on the frame_encoding parameter.
  - **Type**: Any

>   - **Requirements/Conventions**: The value MUST be present.

Reviewer 4: This is not true for the `linear` case (I think it's the only one?), to be fixed.

Reviewer 4: Also, for 'constant', is this a list of 1 value or a single float?

Author: You are right, for linear the value is not needed.

For constant, the format will depend on the property that is shared. It could be a float or a list of one value. But it can also be a list of lists such as for, for example, the lattice vectors.

           },
           "species_at_sites":{
             "frame_serialization_format": "constant",
           },
           "_exmpl_pressure":{
             "frame_serialization_format": "explicit_custom_sparse",
             "nvalues": 20
           }
           "_exmpl_temperature":{
             "frame_serialization_format": "explicit_regular_sparse",
             "step_size_sparse": 10
             "nvalues": 100
           }
         }

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The preceding properties 7.3.1-4 and the fields described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.

The data from the trajectory frames SHOULD only be returned when the user specifically requests these properties in the response_fields.

Next to this the client MAY specify the following parameters to customize the return from the server.
While these URL query parameters are OPTIONAL for clients, API implementations MUST accept and handle them.
The numbering of the frames is zero based, so the first frame is frame number 0.

> - **first_frame**:

Author: At present, only one subrange can be returened at a time. Do we want to support for queries requesting multiple subranges?

Reviewer 4: It would be good to hear from users, but I think a single range per request is enough (if we allow for getting one every N steps between a first and last frame)?

  - A structure having implicit atoms and using assemblies: :val:`["assemblies", "implicit_atoms"]`

Trajectories Entries
--------------------

- **Description**: The :entry:`trajectories` entry point is used to share data from molecular simulations.
  Usually this data will come from molecular dynamics simulations.
  It can however also be used to share data from structures that are related in an other way.
  For example the successive structures from a Monte Carlo simulation.

  Some examples of the data that can be shared are the particle positions, the pressure and the energies.
  :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the following properties: `reference_structure`_, `reference_frame`_, `nframes`_ and `available_properties`_.
  Next to this they can optionally have all the fields of the structures entries as well as relationships and database specific fields.

  The `reference_structure`_ is an example of the kind of structures that are in the trajectory, and it is used to query the trajectory entries in the same way as the structures entries.

>   The data belonging to the frames of the trajectory is only returned when this is specifically requested in the :query-param:`response_fields` parameter.

Author: It is mostly the same as the normal `response_fields` query parameter.

The behaviour here is however slightly different. Normally all the standard OPTIMADE fields are returned. For the positions of the particles in a trajectory you could use the "cartesian_site_positions" property that is already in the standard. This could lead a user to think that when they request a trajectory they would automatically get the particle positions. 
I however chose to have the user request which properties they want to have, so they do not request more data from the server than needed. To make this clear to the user I have placed this line here. It is described more in the section "Retrieving the trajectory data", to which I have made a link.

I can remove it here if you think it is to specific for a general introduction of the trajectory endpoint.

>   The data belonging to the frames of the trajectory is only returned when this is specifically requested in the :query-param:`response_fields` parameter.

Reviewer: OK, I think I understand now, thanks. However, I still think this sentence is superfluous. Section "Entry Listing Endpoints" provides sufficient description on how to say which properties should not appear in the response unless explicitly requested.

>   The data belonging to the frames of the trajectory is only returned when this is specifically requested in the :query-param:`response_fields` parameter.

Author: Ok, I have removed the sentence from this section and changed the surrounding text a bit.

  In this case each property has a dictionary as a value which contains the values of this property and information about which value belongs to which frame.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Author: I have added a link to the "Retrieving the trajectory data" section.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Author: I have changed "describe" to "described". I however do not see anything wrong with the indentation. Could you be more specific ?

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Reviewer: OK, thanks for adding the link. I do not see the problem with indentation anymore. Maybe I was mistaken, sorry.

  Queries on individual frames are currently not supported.

> reference_structure

Author: I think a lot of the specification was first written in Markdown. (Giovanni mentioned this approach to me.) There you need to escape underscores. At first, I used Pycharm to edit the RST file. It has some issues with creating links to other sections when there was an underscore in the name. I therefore removed them. I just noticed there are still some escaped underscores left in the document, though. Either I missed them the first time or they were added in later PR's. Do you want me to remove those as well ?

There is indeed a tilde too many. I have removed it.

> reference_structure

Reviewer: I suggest against bundling unrelated changes in this PR. This way conflicts may emerge, and the diff becomes difficult to eyeball.

> reference_structure

Author: I have readded most of the necessary backslashes. So there should be fewer differences with the original OPTIMADE specification.

> reference_structure

Reviewer: Thanks for de-bundling the changes. I will look at `\_` vs. `_` issue.

~~~~~~~~~~~~~~~~~~~~

- **Description**: This is an example of the structures that can be found in the trajectory.
  It can be used to select trajectories with queries and to give a quick visual impression of the structures in this trajectory.
- **Type**: dictionary
- **Requirements/Conventions**:

  - Each trajectory MUST have a reference_structure.

>   - This reference_structure MAY be one of the frames from the trajectory, in that case the `reference_frame`_ field MUST specify which frame has been used.

Author: > Also, is it meaningful to have the reference structure not taken from the frames of the trajectory?

You could use the average position of the atoms as the reference structure. You may for example want to predict the B-factors of a structure. In that case, one of the results from the simulation would be the average positions of the atoms, which you could use as a reference structure.

> I suggest replacing "reference_structure" with "reference structure" as the specification up to this point does not seem to use terms in such a way.

I have added ":property:" in front of reference_structure to make it more similar to how properties are used elsewhere in the text.

Other than that I do not see how my use is different from for example the `chemical_formula_reduced` in line 1984.
or  `response_fields` in line 1805. > or if they are included also when :query-param:`response_fields` is not present.

>   - Queries on the trajectories MUST be done on the information supplied in the reference_structure when the queried property is in the reference_structure.

Author: You are right, the definition is not very clear about how you should query properties in the reference_structure.

There are two ways:

A: ...\trajectories?filter=nelements=2  
B: ...\trajectories?filter=reference_structure.nelements=2

I think I had method A in mind When I wrote:"Queries on the trajectories MUST be done on the information supplied in the reference_structure when the queried property is in the reference_structure."

But I do not have a strong preference. And method B is perhaps more in line with how we perform queries elsewhere on nested fields.
It should mainly be clear that the query will not be performed on the values in an individual frame.

For now, I will add an example of query A, but I can change it to query B if you prefer.

  This makes it easier for a client to estimate the amount of data a query returns.

  It is up to the server to decide which properties to share and there are no mandatory fields.
  When sharing `cartesian_site_positions`_ the `lattice_vectors`_, `species`_, `dimension_types`_ and `species_at_sites`_ MUST however be shared as well.

- **Type**: dictionary of dictionaries
- **Requirements/Conventions**:

  -   **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  -   **Query**: MUST be a queryable property with support for all mandatory filter features.

- **Sub dictionary fields**

  - **frame_serialization**

    -   **Description**: This property describes how the frames and the returned values of a property are related.
        For each **frame_serialization** method there are additional fields that describe how the values belong to the frames.
        These fields should also be present here.
        A complete description of the **frame_serialization** methods and the fields belonging to these methods can be found in the section: `Return Format for Trajectory Data`_

  - **nvalues**:

>     - **Description**: This field gives the number of values for this property.

Reviewer: What does it mean when `nvalues` is different from the total number of frames?

Author: The value of a property may not have been calculated or stored at every frame/timestep. So the number of available values can be less than the number of frames. By specifying the number of values we allow a client to pre allocate memory to store the data.

Reviewer: OK, thanks. But I see no need for a client to pre-allocate the memory at this point. A client will still read in JSON response with the actual values and do the pre-allocation then by looking at the number of values in the JSON.

Author: The amount of data in a trajectory can be quite large, so the idea was that we would allow streaming JSON parsers to process the files.

Reviewer: Right, but one will always know the top limit, and that limit will be the amount of memory required in the worst-case scenario. I do not think that nowadays one would be unable to pre-allocate the maximum required amount of memory and re-allocate afterwards. Or use dynamic allocation on-the-fly (think linked lists).

Author: The size of a trajectory can definitely be larger than the amount of memory you have available in your machine. 
See for example this [simulation](https://phys.org/news/2019-04-scientists-billion-atom-biomolecular-simulation.html).
Most computers would already struggle to process a single structure of this trajectory.

I don't think OPTIMADE will be used for simulations that large, but the simulation data could still be several gigabytes in size. 
This can be too large for a simple computer to handle. That is part of the reason we allow the selection of a range of frames. That way, the client can request the data in several steps, if it cannot process it in a streaming fashion.
Knowing how many values there are of a property in the trajectory can help with estimating the average amount of data per frame.

    - **Examples**:

      - :val:`100`

- **Examples**:

    .. code:: jsonc

         "available_properties": {
           "cartesian_site_positions": {
             "frame_serialization_format": "explicit",
             "nvalues": 1000
           },
           "lattice_vectors":{
             "frame_serialization_format": "constant",
           },
           "species":{
             "frame_serialization_format": "constant",
           },
           "dimension_types":{
             "frame_serialization_format": "constant",
           },
           "species_at_sites":{
             "frame_serialization_format": "constant",
           },
           "_exmpl_pressure":{
             "frame_serialization_format": "explicit_custom_sparse",
             "nvalues": 20
           }
           "_exmpl_temperature":{
             "frame_serialization_format": "explicit_regular_sparse",
             "step_size_sparse": 10
             "nvalues": 100
           }
         }

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The preceding properties 7.3.1-4 and the fields described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.

Author: Yes, you are right, I was first hoping there was a way to update it automatically like in Latex. But I never found out how to do this in Restructuredtext.

The data from the trajectory frames SHOULD only be returned when the user specifically requests these properties in the response_fields.

Next to this the client MAY specify the following parameters to customize the return from the server.

> While these URL query parameters are OPTIONAL for clients, API implementations MUST accept and handle them.

Author: I have tried to clarify this a bit in the text. I however do not see why a user would misuse these query parameters at the structures endpoint as they do not make sense there.

> While these URL query parameters are OPTIONAL for clients, API implementations MUST accept and handle them.

Reviewer: My comment is more about trying to control the namespace. I would prefer explicitly relating these `*_frame` query parameters to `/trajectories` as in future identically named parameters may appear in other endpoints with completely different meaning. But I am not sure how to properly do so, as we hadn't endpoint-specific query parameters before.

> While these URL query parameters are OPTIONAL for clients, API implementations MUST accept and handle them.

Author: That is an interesting point.
I think this goes beyond the scope of just this PR, as it may also affect other endpoints in the future.
So perhaps it would be good to open a separate issue about this.

It should not be too hard to change the name of a query parameter in the client and server software once that discussion reaches a conclusion. So I think that for now we can continue with finalizing this PR. It will take some time to implement the new trajectory endpoint anyway. So it should be possible to conclude that discussion before trajectory endpoints have been fully implemented.

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The preceding properties `reference_structure`_, `reference_frame`_, `nframes`_, `available_properties`_ and the fields described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.

Reviewer: "fields described" -> "properties described"

Author: I have changed this.

Example of returned trajectory
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is an example of the data field of a JSON object that could be returned after the following query:
:query-url:`http://example.com/optimade/v1/trajectories/traj00000001`

.. code:: jsonc

    "data":{
      "id": "traj00000001",
      "type": "trajectories",

>       "last_modified":"2021-07-16T18:02:03Z"

Reviewer 3: `last_modified` also needs to be demoted to an attribute

Author: Thank you for looking at it again. I have moved it.

          "nelements": 2,
          "elements_ratios": [0.666667,0.333333],
          "chemical_formula_descriptive": "H2O",
          "chemical_formula_reduced": "H2O",
          "chemical_formula_anonymous": "A2B",
          "dimension_types":[0,0,0],
          "nperiodic_dimensions": 0,
          "lattice_vectors" : [[4.0,0.0,0.0],[0.0,4.0,0.0],[0.0,0.0,4.0]],
          "cartesian_site_positions" : [[2.0,2.0,2.0],[1.238,2.0,1.416],[2.762,2.0,1.416]],
          "nsites":3,
          "species_at_sites":["O1","H1","H2"],
          "species":[
            {
              "name":"O1",
              "chemical_symbols":["O"],
              "concentration":[1.0],
            },
            {
              "name":"H1",
              "chemical_symbols":["H"],
              "concentration":[1.0],
            },
            {
              "name":"H2",
              "chemical_symbols":["H"],
              "concentration":[1.0],
            }
          ]
          "relationships": {
            "references": {
              "data": [
                {
                  "type": "references",
                  "id": "dummy/2019"
                }
              ]

>             }

Reviewer 3: `relationships` is not an attribute so should be bumped up

Author: Thank you, I will move it.

          "frame_serialization_format": "constant",
          "values":[[
            {
              "name":"O1",
              "chemical_symbols":["O"],
              "concentration":[1.0]
            },{
              "name":"H1",
              "chemical_symbols":["H"],
              "concentration":[1.0]
            },{
              "name":"H2",
              "chemical_symbols":["H"],
              "concentration":[1.0]
            }
          ]]
        },
        "species_at_sites":{
          "frame_serialization_format": "constant",
          "values":[["O1","H1","H2"]]
        },
        "_exmpl_temperature_set":{
          "frame_serialization_format": "explicit_custom_sparse",
          "sparse_frames":[0,4,5,9],
          "values":[273,273,293,293]
        }
      }
    },
    "links":{
      "next":"http://example.com/optimade/v1/trajectories/traj00000001?response_fields=cartesian_site_positions, lattice_vectors,dimension_types,_exmpl_time,_exmpl_ekin,species,species_at_sites,relationships&first_frame=10",
      "base_url": {
        "href": "http://example.com/optimade",
        "meta": {
          "_exmpl_db_version": "3.2.1"
        }
      }
    }

>     //,...

Reviewer: `//,...` -> `// ...`

  - A structure having implicit atoms and using assemblies: :val:`["assemblies", "implicit_atoms"]`

Trajectories Entries
--------------------

> - **Description**: The :entry:`trajectories` entry is used to share data belonging to sequences of structures, e.g., from a molecular dynamics simulation or a Monte Carlo simulation.

Reviewer 3: ```suggestion
- **Description**: The :entry:`trajectories` entry is used to share data belonging to indexable sequences of structures, e.g., from a molecular dynamics simulation or a Monte Carlo simulation.
```

Not sure if "indexable"/"indexed" is quite what I mean here, but some words that distinguish this field from a collection of structures would be useful, i.e., collections could be used to store a sequence of structures, but there is no mechanism (yet discussed) for slicing/indexing within collections, which I think is the main distinction here (perhaps @rartino has thoughts given his comment on collections elsewhere).

Author: I am not sure what you mean with indexable. It is not my intention to support queries on individual frames.
So they do not need to be indexed by the database backend.

I do not think we should make an artificial distinction between collections and trajectories. In principle, the trajectory endpoint can be used to share collections of structures as well. The use of a reference_structure however only makes sense when the structures have a lot of similarities. If the structures are very diverse, you would not be able to find them efficiently with queries.

Perhaps we could use "sequences of similar structures" but I am not sure whether this would be clear to all readers.

@rartino What do you mean with jumping back and forth? 

In the BioExcel case, we generate the cartesian_site_positions for each of the frames from a list of 32-bit floating point values. We know the number of particles and the number of bytes needed to store the coordinates, so we can set the position of the file pointer to read any frame we want. @rartino is this what you meant with generated on the fly?
They generate some of the files for their files endpoint on the fly as well.

Reviewer 2: With "jumping back and forth", I mean that a server support clients to ask for, e.g., frame 10-30, 100-120, and then 60-80, etc. I'd be very happy to see this functionality as an optional part of the standard; and it is great that you have found a way to efficiently handle these requests for your data set. However, not all providers - including myself - will be able to provide this feature at a reasonable resource usage. For ourselves, my hope is that the standard will allow me to provide our MD run data via the `/trajectory` endpoint metadata, and then providing the trajectory data *only* via a "next"-link paginated unbroken sequence of frames.

Regarding the terminology of sequences vs sets. The design of `/trajectory` is that the frames are ordered using an unbroken sequence of integers. I do not see why we need to relate this to any underlying "physical" relationship between the frames, it is IMO correct to describe the trajectory entry type to be metadata + a sequence of frames.

Author: We have a REST api, so the server does not keep track of what the client has previously requested.
Your server thus MUST be able to handle such requests, assuming that the client requests the ranges(10-30, 100-120, and then 60-80) in separate requests. 

In what kind of format do you store your trajectory data? If it is a fixed width format, you can calculate where in the file a new frame starts.

Reviewer 2: > We have a REST api, so the server does not keep track of what the client has previously requested.

This is frequently not the case in practice for many REST implementations, and when using the approach of giving out a 'handle' that the client uses to represent the state it can be argued it is not against what is meant by the 'statelessness' of REST. Anyone with that handle can make the same continuation request, so the *protocol* is stateless.

> Your server thus MUST be able to handle such requests, assuming that the client requests the ranges(10-30, 100-120, and then 60-80) in separate requests.

I do not see why this has to be a MUST-level requirement. Surely, if we agree on it, we can put it on the level of providing an OPTIMADE paginated response of the frame data - which allows me to just provide a 'next' link that embeds the state handle? I do not think I will be able to deliver random access for our MD data.

> In what kind of format do you store your trajectory data? If it is a fixed width format, you can calculate where in the file a new frame starts.

Most of what we have so far is in bz2:ed VASP XDATCAR files; where my intent was to translate them on-the-fly frame-by-frame in sequence as the data is requested.

Author: > I do not see why this has to be a MUST-level requirement. Surely, if we agree on it, we can put it on the level of providing an OPTIMADE paginated response of the frame data - which allows me to just provide a 'next' link that embeds the state handle? I do not think I will be able to deliver random access for our MD data.

Perhaps a MUST level is too strong and SHOULD, would be a better. 
Returning the whole trajectory instead of only a part of the trajectory would however be quite inefficient, and I do not see any case where it would be impossible to return a subrange of frames.
Making a list of pointers which specify where in the file each frame starts should be possible for nearly all trajectory file formats. 
The ability to return frames from a certain pointer onward would also be needed to handle the next link so that part would not require any extra work.

So I think it would be possible to implement this for most if not all databases. Finally, it would always be possible to convert the data to a data format that is easier to share. 

In case we do make it optional, we should define fields at the info endpoint where a client can check which methods are supported by the server.
 
If a server does not support slicing of the trajectory, you would have to send the whole trajectory. In that case it would be more efficient to send the bzipped trajectory file directly, although it would make processing the data more difficult for the client..

>  Most of what we have so far is in bz2:ed VASP XDATCAR files; where my intent was to translate them on-the-fly frame-by-frame in sequence as the data is requested.

Is this a good example of a [VASP XDATCAR file](https://www.ovito.org/wp-content/uploads/asgarosforum/2882/firstframesXDATCAR.txt) ?
It seems that in that case, each frame has the same size, so you can calculate where the data for a certain frame begins.

The BZIP2 compression makes things more complicated, but not impossible.
A BZIP2 file contains blocks that can be decompressed individually. So it is not necessary to decompress the entire file.
Each block contains a fixed amount of data. So we can calculate which blocks we need to decompress. Unfortunately, the BZIP2 format does not store the size of the compressed blocks, so we will have to make and store a list with the starting points of each block. With this list, we can then determine which blocks we need to decode. 

An alternative approach would be to start decompressing the bzip2 file at approximately the correct location (each block will contain the same type of data, so they will all have approximately the same size.) From that point, you can look for the magic byte sequences that define the start of a block. Once the block is decoded, you can check the value of the "Direct configuration" to see if you are at the correct frame.

Reviewer 2: > Returning the whole trajectory instead of only a part of the trajectory would however be quite inefficient, and I do not see any case where it would be impossible to return a subrange of frames.

The way I see OPTIMADE design, it has never been about what is technically possible/impossible, but that we do not want to put a burden on providers to have to reorganize their data and change the tables/documents/hierarchies in their underlying backend to participate. I feel pretty safe in saying that not everyone with MD data stored today can provide access to individual frames in that data with a reasonable resource usage. 

It is also not clear to me why this is such an important feature that needs to be on MUST level? I've participated in plenty of projects dealing with MD data. In many of them it could have been useful with a standardized format (and a standard communications protocol to transfer the data), but I cannot recall any case where a strong need appeared for transferring partial ranges of MD data. I understand that use cases exist -- e.g., if one wants build a UI on top of OPTIMADE that allows looking at the frames and "scrubbing" the timeline -- so I am very happy this part of the standard exist. However, do we prefer people to not share their MD data at all, rather than share it without random access to frames? Because that is what we are saying with "MUST".

> The ability to return frames from a certain pointer onward would also be needed to handle the next link so that part would not require any extra work.

No. All I need to do is:
- Open a file pointer on the initial request and associate the pointer with an abstract handle id.
- Read as many frames that fits in a paginated response.
- Give a 'next' link that references that handle.
- The next request I receive that has that handle reference, I use the file pointer to transfer the next paginated part of the MD data, until the end is reached.
- I remove expired handles from the dictionary that keeps track of them after a set timeout; requests referencing an unknown/expired handle receives an error indicating that the client has to re-query.

> If a server does not support slicing of the trajectory, you would have to send the whole trajectory. In that case it would be more efficient to send the bzipped trajectory file directly, although it would make processing the data more difficult for the client..

The point for interoperability is to provide data using a carefully standardized format. I don't think bzip2:ed VASP XDATCARs is a good standard.

> Is this a good example of a VASP XDATCAR file ? [...] so we will have to make and store a list with the starting points of each block.

Right, so you end up prescribing that I have to restructure my data representation - in this case build indexes we don't yet have. It isn't that I *cannot* do that - I *can* even just pre-convert all our MD data into an OPTIMADE friendly format and also store them that way. But, it is a highly non-trivial amount of work.

Author: > The point for interoperability is to provide data using a carefully standardized format. I don't think bzip2:ed VASP XDATCARs is a good standard.

Good point.

> I feel pretty safe in saying that not everyone with MD data stored today can provide access to individual frames in that data with a reasonable resource usage.

That's why I wrote:
> Perhaps a MUST level is too strong and SHOULD, would be a better.
Although, I think returning only a part of a trajectory instead of the whole trajectory would be faster for practically all file types.

>No. All I need to do is:
> -  Open a file pointer on the initial request and associate the pointer with an abstract handle id.
> -  Read as many frames that fits in a paginated response.
> -  Give a 'next' link that references that handle.
> -  The next request I receive that has that handle reference, I use the file pointer to transfer the next paginated part of the MD data, until the end is reached.
> -  I remove expired handles from the dictionary that keeps track of them after a set timeout; requests referencing an unknown/expired handle receives an error indicating that the client has to re-query.

I think this approach has quite some similarities with what I suggested. 
In my case, I would store a list with the file pointers at which each frame starts.
In your case, you would store a dictionary with the handle IDs as the keys and the file pointer from which the rest of the trajectory should be read as one of the keys.
So in both cases, we need to be able to read frames from a certain file pointer onward.

Your method has the advantage that:
- You could use a much smaller next link because you could store the rest of the query with the handle. 
- Your dictionary with the handles would be much smaller than the list I proposed.

The disadvantages would be that:
- you would not be able to provide random access.
- If there are multiple threads/servers, you would need to either make sure the next request was executed by the same server/thread or synchronize your dictionary between different servers/threads.

Reviewer 2: Right, so we've settled on not MUST level. I think MAY would be more suitable, but that distinction is not so important.

But, just to clarify something on the point you are making:

> Your dictionary with the handles would be much smaller than the list I proposed.

Indeed - the overhead in keeping track of a list of open file pointers scales with the number of users who simultaneously request data from the server(s) at the cost of a file pointer/user. The overhead of pre-caching pointers to frames scales with the total number of MD frames in my database. In normal circumstances these two approaches have a qualitative difference in resource usage.

Also, you forgot in the list that pointers to frames require significant pre-processing of data before it can be served.

Finally, on this:
> I think returning only a part of a trajectory instead of the whole trajectory would be faster for practically all file types.

Well, I've certainly seen gzip:ed XDATCAR files in the past.

Author: I have changed MUST to SHOULD at line 2982.

>  Well, I've certainly seen gzip:ed XDATCAR files in the past.
It depends on which program was used to generate the gzip files, but often they can indeed only be decoded from the start. 

In principle, it would still be possible to provide access at fixed points in the file by storing the preceding 32kb along with the file pointer. Depending on the size of the frames, this may however be a non-negligible amount of data.  
So it could be impractical to do this for every frame, and such access points could, for example, only be there for one out of every 100 frames.  You would also have to store this 32kb of data, but as you keep track of far fewer points from which the file could be accessed, it would be much less data.

Reviewer 2: To return to the topic of this thread - especially with MUST -> SHOULD, IMO "indexable" should not be part of the definition of a trajectory.

@ml-evs: Isn't "sequence" (which is already there) the word that distinguishes a trajectory from a set? I.e., the fact that the frames are numbered with a contiguous list of integers?

  The individual structures of the trajectories are called frames.

  Some examples of the data that can be shared are the particle positions, the pressure and the energies.
  :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the following properties: `reference_structure`_, `reference_frame`_, `nframes`_ and `available_properties`_.
  Next to this they can optionally have all the fields of the Structures Entries`_ as well as relationships and database specific fields.

  The `reference_structure`_ is an example of the kind of structures that are in the trajectory.
  It is used to query the trajectory entries in the same way as the structures entries.
  For each property in a trajectory there is a dictionary which contains the values of this property in the trajectory and information about which value belongs to which frame.
  It is possible to request only part of a trajectory and to request only 1 out of every n frames.
  This is described in more detail in the section `Retrieving the trajectory data`_.
  Queries on individual frames are not supported.

reference_structure
~~~~~~~~~~~~~~~~~~~

> - **Description**: This is an example of the structures that can be found in the trajectory.
>   It can be used to select trajectories with queries and to give a quick visual impression of the structures in this trajectory.

Reviewer 3: An attempt at rewording to better reflect (my understanding of) the purpose of this field

```suggestion
- **Description**: A representative example of the structures that comprise trajectory.
  This structure provides an overview of the trajectory and is used when filtering trajectories according to fixed properties of the underlying frames, such as composition or dimensionality.
```

Reviewer 2: It has to be "a trajectory" or "the trajectory" on the first line, right? (@ml-evs if you agree - just edit your 'suggested change')

Author: I think "make up" is more common than "comprise" so I'll used that instead. 
I am not sure "overview" is entirely correct word here. To me it suggests some kind of average which is not what this structure is. Therefore I removed "provides an overview of the trajectory and" from Matthews suggestion.

- **Type**: dictionary
- **Requirements/Conventions**:

  - Each trajectory MUST have a :property:`reference_structure`.
  - This :property:`reference_structure` MAY be one of the frames from the trajectory, in that case the `reference_frame`_ field SHOULD specify which frame has been used.

>   - This reference frame has the same properties as the `Structures Entries`_ namely:

Reviewer 3: ```suggestion
  - This reference frame has the same properties outlined in `Structures Entries`_, namely:
```

    - `elements`_
    - `nelements`_
    - `elements_ratios`_
    - `chemical_formula_descriptive`_
    - `chemical_formula_reduced`_
    - `chemical_formula_hill`_
    - `chemical_formula_anonymous`_
    - `dimension_types`_
    - `nperiodic_dimensions`_
    - `lattice_vectors`_
    - `cartesian_site_positions`_
    - `nsites`_
    - `species_at_sites`_
    - `species`_
    - `assemblies`_
    - `structure_features`_

  - The subfields of the :property:`reference_structure` MUST have the same queryability as in the :entry:`structures` entries.

reference_frame
~~~~~~~~~~~~~~~

- **Description**: The number of the frame at which the `reference_structure`_ was taken.
  The first frame is frame 1.
- **Type**: integer
- **Requirements/Conventions**: The value MUST be larger than 0 and equal or less than nframes.

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.

>   - SHOULD NOT be :val:`null` if the `reference_structure`_ is in the trajectory.
>   - MUST be :val:`null` or omitted if `reference_structure`_ is not in the trajectory.

Reviewer 3: ```suggestion
  - SHOULD NOT be :val:`null` if the `reference_structure`_ exactly matches the frame at the index `reference_frame`_ in the trajectory.
  - MUST be :val:`null` or omitted if `reference_structure`_ is not an exact match to any one frame in the trajectory.
```

- **Examples**:

  - :val:`42`

nframes
~~~~~~~

> - **Description**: The number of the frames in the trajectory.

Reviewer 3: ```suggestion
- **Description**: The number of frames in the trajectory.
```

Reviewer 3: More meta question, is this equal to the number of frames in the OPTIMADE representation of the trajectory, or the trajectory as simulated (e.g., `ceil(total_time / time_step)` for MD). The description might have to be tweaked to make either direction clearer.

Author: Good point. I would argue that it is the number of frames exposed by the API. So it would not necessarily match the number of calculation steps used to calculate the trajectory.

  -   **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  -   **Query**: MUST be a queryable property with support for all mandatory filter features.

- **Sub dictionary fields**

  - **frame_serialization_format**

    -   **Description**: This property describes how the frames and the returned values of a property are related.
        For each :property:`frame_serialization_format` method there are additional fields that describe how the values belong to the frames.
        These fields should also be present here.
        A complete description of the :property:`frame_serialization_format` methods and the fields belonging to these methods can be found in the section: `Return Format for Trajectory Data`_

  - **nvalues**:

    - **Description**: This field gives the number of values for this property.
    - **Type**: integer
    - **Requirements/Conventions**: The value MUST be present when :property:`frame_serialization_format` is set to explicit, explicit_regular_sparse or explicit_custom_sparse.
    - **Examples**:

      - :val:`100`

- **Examples**:

    .. code:: jsonc

         "available_properties": [
           {

>              "property": "cartesian_site_positions", 

Reviewer 2: ```suggestion
             "property": "cartesian_site_positions",
```
Commit this to remove the trailing whitespace that currently breaks tests.

Author: Thanks, that seems to have done the trick.

  - A structure having implicit atoms and using assemblies: :val:`["assemblies", "implicit_atoms"]`

Trajectories Entries
--------------------

- **Description**: The :entry:`trajectories` entry point is used to share data from molecular simulations.
  Usually this data will come from molecular dynamics simulations.

>   It can however also be used to share data from structures that are related in an other way.

Reviewer: "an other" -> "another"

Author: I will changed it.

>   For example the successive structures from a Monte Carlo simulation.

Author: I had already changed the text in this section after moving removing the definitions, but I have tried to include the changes suggested here as best as I could.

  Some examples of the data that can be shared are the particle positions, the pressure and the energies.
  :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the following properties: `reference_structure`_, `reference_frame`_, `nframes`_ and `available_properties`_.

>   Next to this they can optionally have all the fields of the structures entries as well as relationships and database specific fields.

Author: I will put this suggestion in the text as well.

  The `reference_structure`_ is an example of the kind of structures that are in the trajectory.
  It is used to query the trajectory entries in the same way as the structures entries.
  For each property in a trajectory there is a dictionary which contains the values of this property in the trajectory and information about which value belongs to which frame.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Author: The checks fail when I directly commit rartino's code suggestions so I will now add these manually.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Reviewer 2: It is the trailing whitespace check that seems to be failing currently. I'll quickly edit my code suggestions to verify that they do not add new trailing whitespace.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Author: I already removed all the trailing white space I could find, but it still gave me an error.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Reviewer 2: There was a whitespace in the `available_properties` example that I removed. I didn't find any other.

>   It is possible to request only part of a trajectory and to request only 1 out of every n frames.

Reviewer 2: I've added as a separate comment (look at the bottom of this PR discussion) the edit that removes the trailing whitespace that presently breaks the tests. The rest of my code suggestions should not break anything.

  This is described in more detail in the section `Retrieving the trajectory data`_.

>   Queries on individual frames are not supported.

Author: Which lines are you referring to ?

>   Queries on individual frames are not supported.

Reviewer 2: All of the marked text - i.e., as it looks now, lines 2833 to 2838.

>   Queries on individual frames are not supported.

Author: This text is part of the general description of the trajectory endpoint, and together with the previous paragraph it gives a summary of what the trajectory endpoint is. 
The information in this section can also be found elsewhere in the text. So I think it can be removed without changing the specification of the trajectory endpoint, although it would make it more difficult to get a good idea of what the trajectory endpoint is with a quick glance.
I'll remove this section, as both you and Merkys seem to be in favour of removing it.

reference_structure
~~~~~~~~~~~~~~~~~~~

- **Description**: This is an example of the structures that can be found in the trajectory.
  It can be used to select trajectories with queries and to give a quick visual impression of the structures in this trajectory.
- **Type**: dictionary
- **Requirements/Conventions**:

  - Each trajectory MUST have a :property:`reference_structure`.

>   - This :property:`reference_structure` MAY be one of the frames from the trajectory, in that case the `reference_frame`_ field MUST specify which frame has been used.

Reviewer 2: Is the MUST requirement on `reference_frame` reasonable? The way I read this, the creator of a synthetic example structure (rather than just taking one out of the trajectory) MUST dig through the data to ensure not to have accidentally recreated an image that appear in the trajectory. I agree that it in many cases this would be unlikely to happen - but since trajectories are for sequences also outside MD, it seems possible and I think it is an unnecessary harsh requirement. I'd argue to lower it to SHOULD.
```suggestion
  - This :property:`reference_structure` MAY be one of the frames from the trajectory, in that case the `reference_frame`_ field SHOULD specify which frame has been used.
```

>   - Queries on the trajectories MUST be done on the information supplied in the :property:`reference_structure` when the queried property is in the :property:`reference_structure`.

Reviewer: To me this seems as an unnecessary deviation from the current specification. We can easily get away without this special additional rule for queries by querying for, e.g., `trajectories?filter=reference_structure.nelements=2`.

Reviewer 2: I do not understand how to read this MUST, what is it meant to enforce?

Author: Reading back this line, it is indeed needlessly restrictive/unclear.

My example Heroku server also supports queries like `cartesian_site_positions.values > 40` and `reference_structure.nsites < 40`

Before I rewrite this; Do we want to keep allowing a query like `nsites < 40`as shorthand for `reference_structure.nsites < 40`, which was what this line was originally meant to describe.

Reviewer: > Before I rewrite this; Do we want to keep allowing a query like `nsites < 40`as shorthand for `reference_structure.nsites < 40`, which was what this line was originally meant to describe.

I would say no. I would feel more comfortable with less deviations from the current filtering idea.

Author: Ok, I have removed these lines.

Author: @rartino Do you mean the must in line  "Queries on the trajectories MUST be done on the information supplied in the :property:`reference_structure` when the queried property is in the :property:`reference_structure`."
or another MUST?

Reviewer 2: Yes, I meant that MUST (sorry for being unclear, it looked more clear when I commented inline in the view differences mode). From your discussion above I gather that it has to do with the PR wanting to enable queries on subfields of `reference_structure` but without the dot notation. However, I'd argue that this is not easily understood from the sentence as phrased now. I also agree with @merkys that it would be better to avoid this somewhat unexpected extension to the protocol.

If `reference_structure.nsites` feels bulky (and I do somewhat agree with that), why not just skip bundling all this inside a `reference_structure` and instead define them as outermost properties, and in each description say that they refer to a reference structure representative of the structures in the trajectory?

Author: The structure properties keys like `nsites` can already be in the outermost properties. E.g. in a grand canonical simulation, the number of particles can change during the simulation. So there would already be a trajectory property 'nsites' holding a dictionary with properties like frame_serialization_format and the values of `nsites` at each of the frames. So it is not possible to place these fields in the outermost level.

Reviewer 2: Oh, ok - I realize from this reply that I had missed how the current design mixes everything - both trajectory metadata and frame data - under "the same entry" in a way this is a bit at odds with the jsonapi design and relies on giving quite a special role to `response_fields`.

I understand that this may have already been discussed at length, and settled, by the people partaking in the design discussions on trajectories on the last meeting. Nevertheless, why did we not settle for a design more in line with the articles and comments examples from JSON API? There, one retrieve data about article 1 as:
  http://example.com/articles/1
and its comments as:
  http://example.com/articles/1/comments

For us, that would suggest a design more in line with, e.g.:
  http://example.com/trajectories/1
and:
  http://example.com/trajectories/1/frames
where one could then specify a frame range using frame properties, e.g.,
  http://example.com/trajectories/1/frames&filter=framenbr>42 and framenbr<42
The response to these request could then be more or less the same as trajectory data example in this PR.

Is there a reason the design isn't like this?

Reviewer: Overall, query support for the reference frame only makes little sense to me. Consider a situation where the reference frame does not match client's filter although there are frames that would. Related to that, will there be any guidelines for reference frame selection?

Maybe it would be better to include ranges (min, max) of each parameter instead, but then the number of parameters is allowed to change if, let's say, new particles appear in some of the frames. Maybe it is then better to support queries on parameters that are guaranteed to stay fixed in a simulation, for example, cell parameters in invariable-cell simulations?

Author: @merkys 
There is no guideline, it is up to the database provider to decide which frame is most representative. 
I thought it would be handy to have a reference structure because it allows making a visualization of the structure in the trajectory.  For most trajectories, only the cartesian_site_positions change, so the reference structure is sufficient for querying.

For those simulations where other properties change, such as the elements that are present or the number of atoms, it is not ideal.
A quick fix would be to add fields to the individual trajectory properties. So we could have properties like: `nsites.average`, `nsites.minimum`, `nsites.maximum` and `elements.set` to allow more flexible querying.
For most of the optimade fields this would work.
For chemical formula's and lattice vectors it is however not immediately obvious how this should be done, as there is not a clear way to define a minimum or a maximum and there are a practically infinite number of possible values.
These fields are however not so interesting for querying. But it does mean we cannot make fields like `.minimum` and `.set` 
MUST fields.

Trajectories can have 100.000's of frames, so selecting trajectories by querying individual frames does not seem practical.

Author: @rartino 
I don't think we discussed this option.
In the current proposal, all the data is contained in one response, so a response would be similar to a file format, which probably makes it a bit easier to process the response.  
(Although large trajectories may need to be split up, as processing large JSON files can be difficult.)

By having a separate endpoint for the frames, I would also expect that the frames behave as objects on their own, just as for example the structures at the structures endpoint. 
So I would expect a return like `data:[{data_frame_1}, {data_frame_2}, {data_frame_3},...]`
Where each data_frame would contain all the data belonging to that frame. 
This would however give a larger response than the data structure in the trajectory proposal and would also be more difficult to generate for the server. Converting a single array to a string can be done fairly fast with NumPy. Generating all the individual frames would however require a lot more looping, which is relatively slow in python.

Reviewer 2: You certainly have a point in that one would expect:

  http://example.com/trajectories/1/frames

to just return a sequence of "frames"-type jsonapi resource objects, and we apparently want to stay away from that for efficiency reasons and instead provide data in separate sequences for everything that can be listed in `available_properties`. It still seems to me as a smaller deviation from the OPTIMADE and jsonapi structure to do something like this:
  http://example.com/trajectories/1/frameset?filter=framenbr>10 and framenbr <100
and allow a frameset to be more of a dynamically generated resource object that represents the data connected to a sequence of frames.

Note that this type of syntax also makes it possible to formulate the type of search that @merkys was asking about via, e.g.:

  http://example.com/trajectories/1/frameset?filter=nsites=42

although, as you say, I would expect few provides to implement this.

I saw that @merkys was actually onto a similar line of thinking with the some kind of sub-encoding construct like `http://example.com/trajectories/1/10-20-100`, although, I think that is further from the jsonapi structure than the above.

Reviewer 2: I've thought some more about this - there really is no way to do it "cleanly" within the jsonapi data model. I seem to recall @ml-evs pointing out at an OPTIMADE web meeting that the key issue is that we do not have a way to handle properties so big that they need pagination. So, the solution is to either (i) not have such data fields (i.e., divide huge entries into sub datatypes like above, e.g., frames / frameranges, etc.) or (ii) to invent a new mechanism to handle such properties. Both (i) and (ii) have their benefits and drawbacks.

The current PR has a design for (ii) that I think is a good solution. However, I also think we should be careful not to end up with different entries (re-)inventing this mechanism in different ways. So, I hope we can settle for semantics that will be somewhat generalizeable. (Which is also the only way we can make this work with the new property data definitions that were recently merged...)

So, looking at how one could think of @JPBergsma's design in more general terms:

- Some properties in an entry are serialized. Perhaps this should be made very obvious, e.g., by a common prefix, i.e., `serialized_cartesian_site_positions`?
- The data in serialized properties should not be returned unless specific REST query parameters relating to serialization are specified. These parameters should then be standardized on OPTIMADE protocol level (rather than for a specific entry); so something like `series_start` and `series_count` or `series_end`, where the only MUST requirement is to accept `series_start=1` to return a paginated series of responses according to what @JPBergsma describes in this PR.
- But then - instead of bundling the serialization info in `available_properties`, it would seem easier to just agree that the serialization info is what is returned for serialized fields when no serialization parameters are specified? E.g., so a response without the `series_start` parameter would return for `serialized_carteasian_site_positions`:
  ```
           {
             "frame_serialization_format": "explicit",
             "nvalues": 1000
           }
  ```
  whereas if one sets `series_start=1`, one gets data in this, and all other, serialized fields.

Author: We can indeed prefix the properties names like `serialized_cartesian_site_positions`.

> - The data in serialized properties should not be returned unless specific REST query parameters relating to serialization are specified. These parameters should then be standardized on OPTIMADE protocol level (rather than for a specific entry); so something like `series_start` and `series_count` or `series_end`, where the only MUST requirement is to accept `series_start=1` to return a paginated series of responses according to what @JPBergsma describes in this PR.

Do I understand correctly that you want to replace `first_frame` with `series_start`, `last_frame` with `series_end` and `frame_step` with `series_count`, so these parameter names are more appropriate for reuse in future endpoints?
I think `series_step` would be a more appropriate replacement for `frame_step`. With `series_count` My first guess would be that it is the number of frames that should be returned.

> But then - instead of bundling the serialization info in available_properties, it would seem easier to just agree that the serialization info is what is returned for serialized fields when no serialization parameters are specified? E.g., so a response without the series_start parameter would return for `serialized_carteasian_site_positions`:
  ```
           {
             "frame_serialization_format": "explicit",
             "nvalues": 1000
           }
  ```

That was what I was thinking about in my post below:
> So far, fields like cartesian_site_positions.frame_serialization_method were present in both the available_properties list and the trajectory property itself. So, I think we can now get rid of this kind of duplication. It also makes this kind of queries a bit simpler. e.g lattice_vectors IS KNOWN vs available_properties.property HAS "lattice_vectors"

Reviewer 2: > Do I understand correctly that you want to replace `first_frame` with `series_start`, `last_frame` with `series_end` and `frame_step` with `series_count`, so these parameter names are more appropriate for reuse in future endpoints?

Right, and the definitions of them being lifted up in the more general parts of the OPTIMADE specification.

My thinking is that `frame` is very trajectory-specific language. I'm far from convinced `series` is the best term either. When I was writing up my reply, I first wrote `sequence_start`, etc., but when I saw that you were already using the term "serialization" for "serialization formats" etc, and went with that for clarity. However, the "thing" that is special with these properties isn't so much that they are "serialized" (everything is serialized when communicated over HTTP), it is that they represent some form of sequential data, I think. (We were indeed onto this in another discussion spawned by @ml-evs.)

> I think series_step would be a more appropriate replacement for frame_step. With series_count My first guess would be that it is the number of frames that should be returned.

Fair point about wanting an (optional) `series_step` or equivalent - it should also be included. I did actually mean `series_count` as you interpreted it, and meant to reflect on whether it would be better (more generalizeable) to give a count rather than the end value. Why did you end up with `frame_end` rather than `frame_count` in your design? 

This is, however, a side point. I think it is best if we settle for a single "magical" query parameter that enable output of serialized/sequenced data; which then should be `series_start`. Hence, it probably should be a client error to specify `series_step` without `series_start`.

Author: We could make `series_start` the "magical" parameter, although the presence of any of `series_start`, `series_count` `series_end` or `series_step`, would indicate that the user wants trajectory information.

It was more than a year ago that we discussed this, so I am not a 100% sure what the reasons for choosing to specify the end point over specifying the number of frames that needed to be returned were. It may simply have been that in python, you can do array indexing as `List[start: end(exclusive): step_size]`. So it seemed logical to do this here in the same way, as this would be the easiest to implement in python.

Reviewer 2: > We could make `series_start` the "magical" parameter, although the presence of any of `series_start`, `series_count` `series_end` or `series_step`, would indicate that the user wants trajectory information.

I do not have a strong opinion; we could indeed say that all of these are "magical". The relatively weak argument to specify a single one (i.e., `series_start`) is that I think the protocol becomes slightly more robust to change. If we later standardize yet another parameter - or if someone makes up their own, e.g., `_exmpl_series_blockify`. If the specification clearly says "Serialized data is returned if, and only if, the query parameter `series_start` is included with a non-empty value in the request." then clients won't expect that just including, e.g., `?_exmpl_series_blockify=1` will return serialized data (which a server that do not recognize that parameter couldn't understand.) 

> It may simply have been that in python, you can do array indexing as `List[start: end(exclusive): step_size]`. So it seemed logical to do this here in the same way, as this would be the easiest to implement in python.

Not very important at all, but my gut feeling is that start + count is slightly more common in similar interfaces for some minor reasons, like: "start from frame 5 and give me 5 frames" unambiguously returns frames 5, 6, 7, 8, 9. For "start from frame 5 and end at frame 10" I'd have to check documentation on whether frame 10 will be returned or not. But if anyone has a slight inclination towards start + end, it would be fine.

Author: In some cases, the query would be shorter because one of the other parameters e.g.(series_end or series_count) would have been specified already. Other than that, there is no real advantage to allowing multiple parameters to trigger sending the values belonging to the trajectory. Although I do like the symmetry of having all these parameters behave in the same way.

We used the terms first and last to make it a bit clearer that it is an inclusive range. At least, I interpret "last" as inclusive.
The number of values returned may be less than `series_count` because not all properties have values defined for each frame. Perhaps this is confusing to some.
 
But I do not have a strong preference for either of these two discussion points.

>     The subfields of the reference_structure MUST have the same queryability as in the `structures entries`_.

Reviewer: Markup seems off. `reference_structure` should be marked with `:property:`, but I cannot figure out what `structures entries` refer to.

Author: I have added the `:property:` flag.
`structures entries`_ is a link to section 8.2 
It is not possible for something to marked and be a link at the same time, so I have now made the first `structures entries` a link and placed the `:entry:` marker in front of the other.
I also moved this line after the list of reference_structure properties, as this seems more appropriate.

Reviewer: Right, but shouldn't it then be "Structures Entries" (capitalized)?

Author: I have capitalized "Structures Entries" when referring to section 8.2

>     For example, the query : http://example.com/optimade/v1/trajectories?filter=nelements=2 would use the `nelements`_ property within the reference_structure.

Reviewer 2: ```suggestion
    For example, the query: http://example.com/optimade/v1/trajectories?filter=nelements=2 would use the `nelements`_ property within the reference_structure.
```

Author: This line is no longer there.

  - This reference frame has the same properties as the structure entries namely:

    - `elements`_
    - `nelements`_
    - `elements_ratios`_
    - `chemical_formula_descriptive`_
    - `chemical_formula_reduced`_
    - `chemical_formula_hill`_
    - `chemical_formula_anonymous`_
    - `dimension_types`_
    - `nperiodic_dimensions`_
    - `lattice_vectors`_
    - `cartesian_site_positions`_
    - `nsites`_
    - `species_at_sites`_
    - `species`_
    - `assemblies`_
    - `structure_features`_

reference_frame
~~~~~~~~~~~~~~~

- **Description**: The number of the frame at which the `reference_structure`_ was taken.
  The first frame is frame 1.
- **Type**: integer
- **Requirements/Conventions**: The value MUST be larger than 0 and equal or less than nframes.

>   - **Support**: MUST be supported if the `reference_structure`_ is taken from the trajectory.
>     If the `reference_structure`_ is not in the trajectory, the value MUST NOT be present.
>   - **Query**: Support for queries on this property is OPTIONAL.
>     If supported, filters MAY support only a subset of comparison operators.
> 

Reviewer 2: This is tricky, but especially in light of the property standardization work, I do not think we want non-standard custom-written "Support"-levels. So, I think the more intricate requirement in the relationship with the trajectory should go in a separate Requirements/Conventions point. This is then what I would suggest if you accept lowering the requirement to SHOULD level:
```suggestion
  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.
  - SHOULD NOT be :val:`null` if the `reference_structure`_ is in the trajectory. 
  - MUST be :val:`null` or omitted if `reference_structure`_ is not in the trajectory.
```

- **Examples**:

  - :val:`42`

nframes
~~~~~~~

- **Description**: The number of the frames in the trajectory.
- **Type**: integer
- **Requirements/Conventions**:

  - **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  - **Query**: MUST be a queryable property with support for all mandatory filter features.
  - The integer value MUST be equal to the length of the trajectory, that is, the number of frames.
  - The integer MUST be a positive non-zero value.

- **Querying**:

  - A filter that matches trajectories that have exactly 100 frames:
    - :filter:`nframes=100`
  - A filter that matches trajectories that have between 100 and 1000 frames:
    - :filter:`nframes>=100 AND nframes<=1000`

- **Examples**:

  -   :val:`42`

available_properties
~~~~~~~~~~~~~~~~~~~~

- **Description**: A dictionary with an entry for each of the properties for which data is available in the trajectory.
  The key is the name of the property.
  The value is a dictionary containing information about which value belongs to which frame.

>   This makes it easier for a client to estimate the amount of data a query returns.

Reviewer 2: The design of `available_properties` breaks something we decided way back: to avoid storing data as keys in  dictionaries. Instead - see e.g. the 'species' property - it was found for several reasons to be a better idea to turn these designs into lists of dictionaries where the key is instead inserted among the right-hand-side values with a sensible field name. I'll edit an example below to show how.

If the definition of available_properties is changed this way, an added advantage should be that it is easier to format this definition of the property the same way as, e.g., `species` where all the type info goes into type. Right now the "Sub dictionary fields" category is something that was introduced here and is used nowhere else.

Author: I am starting to wonder whether we really need the available properties field. 
If we no longer support redirecting queries to the reference structure we could perform queries directly on the properties. 
Instead of only making the properties available when specifically requested in the response field, We could do this just for the values subfield. In that case we could get rid of the available properties field all together.

Reviewer 2: `available_properties` made sense to me... It allows me to query trajectories that have certain types of data, e.g., "give me all trajectories with force data". It also doubles by describing the underlying serialization model for that data. This I would probably never query on - but is good to be able to know what to expect (e.g. to prepare data structures) upon retrieving trajectory data.

Author: It is indeed important to be able to check for which fields data is present in the trajectory. 

If we do not redirect queries to the `reference_structure`, we can however perform queries directly on the fields in the trajectory. So you could do a query like: `filter=lattice_vectors IS KNOWN AND cartesian_site_positions.nvalues > 1`
to find a structure for which the lattice vectors are defined and for which there is more than 1 frame for which the particle positions have been recorded.
So far, fields like `cartesian_site_positions.frame_serialization_method` were present in both the `available_properties` list and the trajectory property itself. So, I think we can now get rid of this kind of duplication. It also makes this kind of queries a bit simpler. e.g `lattice_vectors IS KNOWN` vs `available_properties.property HAS  "lattice_vectors"`

We however, do have to use something like `response_fields=cartesian_site_positions.values` to request the values that are not shared by default.

- **Type**: dictionary of dictionaries
- **Requirements/Conventions**:

  -   **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  -   **Query**: MUST be a queryable property with support for all mandatory filter features.

- **Sub dictionary fields**

  - **frame_serialization_format**

    -   **Description**: This property describes how the frames and the returned values of a property are related.
        For each :property:`frame_serialization_format` method there are additional fields that describe how the values belong to the frames.
        These fields should also be present here.
        A complete description of the :property:`frame_serialization_format` methods and the fields belonging to these methods can be found in the section: `Return Format for Trajectory Data`_

  - **nvalues**:

    - **Description**: This field gives the number of values for this property.
    - **Type**: integer
    - **Requirements/Conventions**: The value MUST be present when :property:`frame_serialization_format` is set to explicit, explicit_regular_sparse or explicit_custom_sparse.
    - **Examples**:

      - :val:`100`

> 

Reviewer: I suggest avoiding surplus empty lines, in consistence with the remaining specification. Single empty line is enough.

- **Examples**:

    .. code:: jsonc

>          "available_properties": {
>            "cartesian_site_positions": {
>              "frame_serialization_format": "explicit",
>              "nvalues": 1000
>            },
>            "lattice_vectors":{
>              "frame_serialization_format": "constant",
>            },
>            "species":{
>              "frame_serialization_format": "constant",
>            },
>            "dimension_types":{
>              "frame_serialization_format": "constant",
>            },
>            "species_at_sites":{
>              "frame_serialization_format": "constant",
>            },
>            "_exmpl_pressure":{
>              "frame_serialization_format": "explicit_custom_sparse",
>              "nvalues": 20
>            },
>            "_exmpl_temperature":{
>              "frame_serialization_format": "explicit_regular_sparse",
>              "step_size_sparse": 10
>              "nvalues": 100
>            }
>          }

Reviewer 2: In line with my comment above (but needs other changes above if accepted), this is how I think this data structure should look:
```suggestion
         "available_properties": [
           {
             "property": "cartesian_site_positions", 
             "frame_serialization_format": "explicit",
             "nvalues": 1000
           },
           {
             "property": "lattice_vectors",
             "frame_serialization_format": "constant",
           },
           {
             "property": "species",
             "frame_serialization_format": "constant",
           },
           {
             "property": "dimension_types",
             "frame_serialization_format": "constant",
           },
           {
             "property": "species_at_sites",
             "frame_serialization_format": "constant",
           },
           {
             "property": "_exmpl_pressure",
             "frame_serialization_format": "explicit_custom_sparse",
             "nvalues": 20
           },
           {
             "property": "_exmpl_temperature",
             "frame_serialization_format": "explicit_regular_sparse",
             "step_size_sparse": 10
             "nvalues": 100
           }
         ]
```

Retrieving the trajectory data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The preceding properties `reference_structure`_, `reference_frame`_, `nframes`_, `available_properties`_ and the properties described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.

Reviewer 2: I agree with @merkys in the closed comment that we at least must inject a "for queries on the trajectory endpoint" here; but there is a more major issue that we don't have a place for these kind of endpoint requirements. To not hold this PR, I think we just need to add an issue to address that on a larger scale. But fore now I suggest:
```suggestion
For queries referencing the :endpoint:`trajectories` endpoint, the preceding properties `reference_structure`_, `reference_frame`_, `nframes`_, `available_properties`_ and the properties described under `Properties Used by Multiple Entry Types`_ MUST be returned when no :query-param:`response_fields` property (see the section `Entry Listing URL Query Parameters`_) is specified.
```

The data from the trajectory frames SHOULD only be returned when the user specifically requests these properties in the response_fields.

> Next to this the client MAY specify the following parameters to customize the return from the server at the trajectory endpoint.

Reviewer 2: ```suggestion
Furthermore, the client MAY specify the following parameters to customize the return from the server at the trajectory endpoint.
```

            "frame_serialization_format": "explicit",
            "nvalues":360
          },
          "species":{
            "frame_serialization_format": "constant",
          },
          "dimension_types":{
            "frame_serialization_format": "constant",
          },
          "lattice_vectors":{
            "frame_serialization_format": "constant",
          },
          "species_at_sites":{
            "frame_serialization_format": "constant",
          },
          "_exmpl_temperature_set":{
            "frame_serialization_format": "explicit_custom_sparse",
            "nvalues":144,
          },
          "_exmpl_time":{
            "frame_serialization_format": "linear",
            "offset_linear": 0,
            "step_size_linear": 1.5
          },
          "_exmpl_ekin":{
            "frame_serialization_format": "explicit_regular_sparse",
            "step_size_sparse": 2,
            "nvalues":180,
          }
        }
        "relationships": {
          "references": {
            "data": [
              {
                "type": "references",
                "id": "dummy/2019"
              }
            ]
          }
        }
      },
    },
    ...

> After the previous querry is an example of a JSON object that could be returned after the following query:

Reviewer 2: "Query" is misspelled, but I also have difficulty parsing the meaning of the rest of the sentence. Is this what is meant?
```suggestion
Upon having performed the above query, the following query can be performed to return trajectory data:
```
(I also removed the mentioning of JSON. I assume I can be allowed to encode the response in `response_format=_httk_hdf5` just like with everything else in OPTIMADE...)

:query-url:`http://example.com/optimade/v1/trajectories/traj00000001?response_fields=cartesian_site_positions, lattice_vectors,dimension_types,_exmpl_time,_exmpl_ekin,species,species_at_sites,_exmpl_temperature_set&first_frame=0`

.. code:: jsonc

    "data":{
      "id": "traj00000001",
      "type": "trajectories",
      "attributes:{

>         "last_modified":"2021-07-16T18:02:03Z"

Reviewer: Missing trailing comma, invalid JSON. Please make sure this example is valid JSON, I have not checked for other possible syntax mistakes.

Author: I have added a few missing quote marks and added and removed commas where necessary.

I now use ellipses to indicate that some of the JSON response is missing. 
To make it correct JSON, I would also have to add other fields such as the meta and links fields. 
Do you want me to do this as well ?

Reviewer: There are ellipses in JSONs in the specification already. They are placed in comments (`// ....`) to render correct JSONC (a.k.a. JSON with comments). Please do the same here.

  The individual steps of the trajectories are called frames.

  Some examples of the data that can be shared are the particle positions, the pressure and the energies.
  :entry:`trajectories` entries have the properties described in the section `Properties Used by Multiple Entry Types`_ as well as the property `nframes`_ and `reference_frame`_.
  Furthermore, :entry:`trajectories` can optionally have relationships, database specific fields and all the properties defined for the structures endpoint.
  If one of these properties does not have a constant value it should be shared as a ranged property, as described in #TODO insert reference to ranged properties section.
  The dimension that corresponds to the steps of the trajectory MUST have :field:`range_id` = :val:`"frames"`.

reference_frame
~~~~~~~~~~~~~~~

- **Description**: The indexes of a set of frames that give a good but brief overview of the trajectory.
  The first frame could for example be a starting configuration, the second a transition state and the third the final state.
- **Type**: list of integers

> - **Requirements/Conventions**: The values MUST be larger than 0 and equal or less than nframes.

Reviewer 5: ```suggestion
- **Requirements/Conventions**: The values MUST be larger than 0 and equal to or less than nframes.
```

Author: During the optimade meeting we decided to start counting from 0 instead, so this remark is no longer valid.

  The dimension that corresponds to the steps of the trajectory MUST have :field:`range_id` = :val:`"frames"` for that dimension in its property definition, See `Property Definition keys from JSON Schema`_.

nframes
~~~~~~~

- **Description**: The number of frames in the trajectory as exposed by the API.
  This value may deviate from the number of steps used to calculate the trajectory.
  E.g., for a 10 ps simulation with calculation steps of 1 fs where data is stored once every 50 fs, nframes = 200.
- **Type**: integer
- **Requirements/Conventions**:

  - **Support**: MUST be supported by all implementations, i.e., MUST NOT be :val:`null`.
  - **Query**: MUST be a queryable property with support for all mandatory filter features.
  - The integer value MUST be equal to the length of the trajectory, that is, the number of frames.
  - The integer MUST be a positive non-zero value.

- **Querying**:

  - A filter that matches trajectories that have exactly 100 frames:
    - :filter:`nframes=100`
  - A filter that matches trajectories that have between 100 and 1000 frames:
    - :filter:`nframes>=100 AND nframes<=1000`

- **Examples**:

  -   :val:`42`

reference_frames

> ~~~~~~~~~~~~~~~

Reviewer 5: ```suggestion
~~~~~~~~~~~~~~~~
```

:query-url:`http://example.com/optimade/v1/trajectories/traj00000001`

.. code:: jsonc

  {
    "data":{
      "id": "traj00000001",
      "type": "trajectories",
      "attributes": {
        "last_modified":"2021-07-16T18:02:03Z",
        "elements": [["H","O"]],
        "nelements": [2],
        "elements_ratios": [[0.666667,0.333333]],
        "chemical_formula_descriptive": ["H2O"],
        "chemical_formula_reduced": ["H2O"],
        "chemical_formula_anonymous": ["A2B"],
        "dimension_types":[[0,0,0]],
        "nperiodic_dimensions": [0],
        "lattice_vectors" : [[[4.0,0.0,0.0],[0.0,4.0,0.0],[0.0,0.0,4.0]]],
        "cartesian_site_positions" : null,
        "nsites":[3],
        "species_at_sites":[["O1","H1","H2"]],
        "species":[[
          {
            "name":"O1",
            "chemical_symbols":["O"],
            "concentration":[1.0]
          },
          {
            "name":"H1",
            "chemical_symbols":["H"],
            "concentration":[1.0]
          },
          {
            "name":"H2",
            "chemical_symbols":["H"],
            "concentration":[1.0]
          }
        ]],
        "reference_frames": [1],

>         "nframes": [360],

Reviewer 6: ```suggestion
        "nframes": 360,
```

        "_exmpl_temperature": null,
        "_exmpl_ekin": null
      },
      "meta":{
        "property_metadata":{
          "cartesian_site_positions":{
            "range":{
              "indexable_dim": ["frames"],
              "data_range": [{
                  "start": 1,
                  "step": 1,
                  "stop": 360,
                },{
                  "start": 1,
                  "step": 1,
                  "stop": 3,
                },{
                  "start": 1,
                  "step": 1,
                  "stop": 3,
                }],
              "layout":"dense",
              "nvalues": 3240
            },
          },
          "_exmpl_temperature":{
            "range": {
              "nvalues": 144,
              "indexable_dim": ["frames"],
              "data_range": [
                {

>                   "start": 1,
>                   "stop": 360,
>                   "step": null
>                 }
>               ],
>               "layout": "sparse"
>             }
>           },
>           "_exmpl_ekin":{
>             "range": {
>               "nvalues": 180,
>               "indexable_dim": ["frames"],
>               "data_range": [
>                 {
>                   "start": 1,
>                   "stop": 360,
>                   "step": 2
>                 }
>               ],
>               "layout":"dense",
>             }
>           }
>         },
>         "partial_data_links": {
>           "cartesian_site_positions": [
>             {
>               "format": "jsonlines",
>               "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/cartesian_site_positions/jsonlines"
>             },{
>               "format": "_exmpl_xyz",
>               "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/cartesian_site_positions/xyz"
>             },
>           ],
>           "_exmpl_temperature": [
>             {
>               "format": "jsonlines",
>               "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/temperature/jsonlines"
>             },
>           ],
>           "_exmpl_ekin": [
>             {
>               "format": "jsonlines",
>               "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/ekin/jsonlines"
>             },
>           ]
>         }
>       },
>       "relationships": {
>         "references": {
>           "data": [
>             {
>               "type": "references",
>               "id": "dummy/2019"
>             }
>           ]
>         }
>       }
>     }
>     //...
>   }

Reviewer 5: ```suggestion
  {
    "data":{
      "id": "traj00000001",
      "type": "trajectories",
      "attributes": {
        "last_modified":"2021-07-16T18:02:03Z",
        "elements": [["H","O"]],
        "nelements": [2],
        "elements_ratios": [[0.666667,0.333333]],
        "chemical_formula_descriptive": ["H2O"],
        "chemical_formula_reduced": ["H2O"],
        "chemical_formula_anonymous": ["A2B"],
        "dimension_types":[[0,0,0]],
        "nperiodic_dimensions": [0],
        "lattice_vectors" : [[[4.0,0.0,0.0],[0.0,4.0,0.0],[0.0,0.0,4.0]]],
        "cartesian_site_positions" : null,
        "nsites":[3],
        "species_at_sites":[["O1","H1","H2"]],
        "species":[[
          {
            "name":"O1",
            "chemical_symbols":["O"],
            "concentration":[1.0]
          },
          {
            "name":"H1",
            "chemical_symbols":["H"],
            "concentration":[1.0]
          },
          {
            "name":"H2",
            "chemical_symbols":["H"],
            "concentration":[1.0]
          }
        ]],
        "reference_frames": [1],
        "nframes": [360],
        "_exmpl_temperature": null,
        "_exmpl_ekin": null
      },
      "meta":{
        "property_metadata":{
          "cartesian_site_positions":{
            "range":{
              "indexable_dim": ["frames"],
              "data_range": [{
                  "start": 1,
                  "step": 1,
                  "stop": 360
                },{
                  "start": 1,
                  "step": 1,
                  "stop": 3
                },{
                  "start": 1,
                  "step": 1,
                  "stop": 3
                }],
              "layout":"dense",
              "nvalues": 3240
            }
          },
          "_exmpl_temperature":{
            "range": {
              "nvalues": 144,
              "indexable_dim": ["frames"],
              "data_range": [
                {
                  "start": 1,
                  "stop": 360,
                  "step": null
                }
              ],
              "layout": "sparse"
            }
          },
          "_exmpl_ekin":{
            "range": {
              "nvalues": 180,
              "indexable_dim": ["frames"],
              "data_range": [
                {
                  "start": 1,
                  "stop": 360,
                  "step": 2
                }
              ],
              "layout":"dense"
            }
          }
        },
        "partial_data_links": {
          "cartesian_site_positions": [
            {
              "format": "jsonlines",
              "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/cartesian_site_positions/jsonlines"
            },{
              "format": "_exmpl_xyz",
              "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/cartesian_site_positions/xyz"
            }
          ],
          "_exmpl_temperature": [
            {
              "format": "jsonlines",
              "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/temperature/jsonlines"
            }
          ],
          "_exmpl_ekin": [
            {
              "format": "jsonlines",
              "link": "https://example.org/optimade/v1.2/extensions/partial_data/trajectories/traj00000001/ekin/jsonlines"
            }
          ]
        }
      },
      "relationships": {
        "references": {
          "data": [
            {
              "type": "references",
              "id": "dummy/2019"
            }
          ]
        }
      }
    }
    //...
  }
```

Reviewer 4: For simplicity I'm merging this without carefully checking, since anyway the example needs to be adapted afterwards
